dnsapi/dns_kas.sh

```diff
@@ -0,0 +1,153 @@
+  _txt_value=$2
+  _info "Using DNS-01 All-inkl/Kasserver hook"
+  _info "Cleaning up after All-inkl/Kasserver hook"
+  _info "Removing $_full_domain DNS TXT entry on All-inkl/Kasserver"
+
+  _check_and_save
+  _get_zone "$_full_domain"
+  _get_record_name "$_full_domain"
+  _get_record_id
+
+  # If there is a record_id, delete the entry
+  if [ -n "$_record_id" ]; then
+    params="?kas_login=$KAS_Login"
+    params="$params&kas_auth_type=$KAS_Authtype"
+    params="$params&kas_auth_data=$KAS_Authdata"
+    params="$params&kas_action=delete_dns_settings"
+    params="$params&var1=record_id"
+    params="$params&wert1=$_record_id"
+    response="$(_get "$KAS_Api$params")"
+    _debug2 "response" "$response"
+    if ! _contains "$response" "TRUE"; then
+      _err "Either the txt record is not found or another error occurred, please check manually."
+      return 1
+    fi
+  else # Cannot delete or unkown error
+    _err "No record_id found that can be deleted. Please check manually."
+    return 1
+  fi
+
+ return 0
+}
+
+########################## PRIVATE FUNCTIONS ###########################
+
+# Checks for the ENV variables and saves them
+_check_and_save() {
+  KAS_Login="${KAS_Login:-$(_readaccountconf_mutable KAS_Login)}"
+  KAS_Authtype="${KAS_Authtype:-$(_readaccountconf_mutable KAS_Authtype)}"
+  KAS_Authdata="${KAS_Authdata:-$(_readaccountconf_mutable KAS_Authdata)}"
+
+  if [ -z "$KAS_Login" ] || [ -z "$KAS_Authtype" ] || [ -z "$KAS_Authdata" ]; then
+    KAS_Login=
+    KAS_Authtype=
+    KAS_Authdata=
+    _err "No auth details provided. Please set user credentials using the \$KAS_Login, \$KAS_Authtype, and \$KAS_Authdata environment variables."
+    return 1
+  fi
+  _saveaccountconf_mutable KAS_Login "$KAS_Login"
+  _saveaccountconf_mutable KAS_Authtype "$KAS_Authtype"
+  _saveaccountconf_mutable KAS_Authdata "$KAS_Authdata"
+  return 0
+}
+
+# Gets back the base domain/zone.
+# TODO Get a list of all possible root zones and compare (Currently not possible via provider)
+# See: https://github.com/Neilpang/acme.sh/wiki/DNS-API-Dev-Guide
+_get_zone() {
+  _zone=$(echo "$1" | rev | cut -d . -f1-2 | rev).
+}
+
+# Removes the domain/subdomain from the entry since kasserver
+# cannot handle _full_domain
+# TODO Get a list of all possible root zones and compare (Currently not possible via provider)
+# See: https://github.com/Neilpang/acme.sh/wiki/DNS-API-Dev-Guide
+_get_record_name() {
+  _record_name=$(echo "$1" | rev | cut -d"." -f3- | rev)
+}
+
+# Retrieve the DNS record ID
+_get_record_id() {
+  params="?kas_login=$KAS_Login"
+  params="$params&kas_auth_type=$KAS_Authtype"
+  params="$params&kas_auth_data=$KAS_Authdata"
+  params="$params&kas_action=get_dns_settings"
+  params="$params&var1=zone_host"
+  params="$params&wert1=$_zone"
+
+  response="$(_get "$KAS_Api$params")"
+  _debug2 "response" "$response"
+
+  _record_id="$(echo "$response" | grep -A 4  "$_record_name" | grep "record_id" | cut -f2 -d">" | xargs)"
```

Comment:
@Neilpang I think without using `grep -A` I can not do it actually. To me this seems to be the only most possible sane solution here, since what I get back is some HTML chunk with some plaintext array at the end, see example: https://gist.github.com/anonymous/eec9b149ab4b5ab8c30541a15e587d80

They also offer a `WSDL`  https://kasapi.kasserver.com/dokumentation/?open=quellcode endpoint that can be accessed via `SOAP` and I think there I would get `XML` back, but I have no clue how to do this without using `curl` actually.

If you have an idea on how to solve this, let me know. thanks!

Reply:
no,  we have removed a lot of `grep -A` before.
show me your `$response`,  and what you want to grep.

Reply:
Hi @Neilpang see the example gist above. I post the example here again:

```
....
            <td align='right'>
               10 <input name='var10' size='25' value=''>
            </td>
            <td>
               <input name='wert10' size='25' value=''>
            </td>
          </tr>
                  <tr>
            <th colspan="2">
               <a target="_blank" href="/dokumentation/?open=quellcode">
                  Beispiel Quellcode ansehen
               </a>
            </th>
         </tr>
         <tr>
            <th colspan="2" align="right">
               Felder:
			   <input size="2" name="anz_var" value="10" />
			   <input type="submit" />
            </th>
            </tr>
      </table>
   </form>
   </div>
   <div style="float:left;">
			   <div class="erfolg">
			  <b>ReturnString: </b>TRUE</br>
			  <b>ReturnInfo: </b><pre>Array
(
    [0] => Array
        (
            [record_zone] => example.com
            [record_name] => 
            [record_type] => NS
            [record_data] => ns6.kasserver.com.
            [record_aux] => 0
            [record_id] => 0
            [record_changeable] => N
        )

    [1] => Array
        (
            [record_zone] => example.com
            [record_name] => *
            [record_type] => A
            [record_data] => 85.13.yyy.xx
            [record_aux] => 0
            [record_id] => 0
            [record_changeable] => N
        )

    [2] => Array
        (
            [record_zone] => example.com
            [record_name] => _acme-challenge.test51
            [record_type] => TXT
            [record_data] => jU6yvYv2Y4rHk2uFNwjxH1gcHmU7DRp4wvjoR6TEfgI
            [record_aux] => 0
            [record_id] => 36986630
            [record_changeable] => Y
        )
)
```

Reply:
@Neilpang have you had already a chance to have a look at it? thanks!

Reply:
Is this still in developement?
I would need this to migrate to acme.sh but it seems like there is no progress for a month now.

Reply:
sorry, for the delay. I was totally missing this PR.

This can grep the correct record id:

```
echo "$response"  | tr -d "\n\r" | sed "s/=> Array/\n=> Array/g" | tr -d " " | tr '[]' '<>' | grep "=>_acme-challenge.test51<" | grep '>TXT<' | tr '<' '\n' | grep record_id | cut -d '>' -f 3

```

Reply:
Thanks @Neilpang! Also bit busy these days and need to see if I can have a look at this again. @dojo90 fyi! If you have time feel free to send me push request!

Reply:
Actually I'm confused. As you (@Neilpang) write here 

> do not use `grep -A` and `xargs`
https://github.com/Neilpang/acme.sh/pull/1341#discussion_r173605438

you give us a solution which uses `grep` 🤔

So I think we can use the actual implementation (https://github.com/phlegx/acme.sh/blob/all-inkl-kasserver-dns-script/dnsapi/dns_kas.sh#L153) ?

Reply:
@dojo90

I mean don't use `-A` option with `grep`.

you can use simple `grep` command.